### PR TITLE
Support Elasticsearch7 along with ES2

### PR DESCRIPTION
**Pull request by sravfeyn**
This PR adds support for ES 7 version while also supporting ES2.

The bulk of changes are to fix breaking changes. Below are ES docs
[Breaking changes in ES5](Breaking changes in ES5
Breaking changes in ES7)
[Breaking changes in ES7](https://www.elastic.co/guide/en/elasticsearch/reference/7.6/breaking-changes-7.0.html)

- Changes to base query structure
- Support ESQuery base class to generate ES version specific queries that powers all ES accessors such as FormES, CaseES etc.
- Changes in mapping depending on ES version
- Ton of changes in test cases that add ES7 queries.
- Some amount of refactors

One note on unit-testing is, for all tests that I think touch ES, I have added a nose attr called es_test. And then, I created a new test matrix that tests all these tests on another ES version than the default. This PR runs all tests on default ES2 and the extra matrix tests es_test tests on ES7. Just to be safe, I also have a build-only PR that does the mirror of this PR, that is, default all tests are run on ES7 and extra matrix to run on ES2.

@snopoke @dannyroberts @esoergel @proteusvacuum @millerdev

Review tips: 
1. Please review all the Sub PRs that are simply refactors/fixes that don't need core ES7 changes.
  https://github.com/dimagi/commcare-hq/pull/28263
  https://github.com/dimagi/commcare-hq/pull/28264
  https://github.com/dimagi/commcare-hq/pull/28265
  https://github.com/dimagi/commcare-hq/pull/28266
  https://github.com/dimagi/commcare-hq/pull/28267
  https://github.com/dimagi/commcare-hq/pull/28268
  https://github.com/dimagi/commcare-hq/pull/28269
2. I rebased about 130 commits into 15 logical commits that you can review one by one. Above 7 PRs correspond to about 7 of these 15 commits.

### CODE REVIEW

**Review by millerdev — COMMENTED** [review]
> Above 7 PRs correspond to about 7 of these 15 commits.

It is pretty hard for a reviewer to figure out which commits have been covered by other PRs and which ones are new material. Please rebase this PR on master after those 7 PRs have been merged and eliminate the changes in this PR that were done in those other PRs.

File: .travis.yml
Comment (sravfeyn) [reply .travis.yml:38]: What's the reason not to make this default?

Comment (snopoke) [reply .travis.yml:38]: it makes the build run for longer when it's going to fail anyway

File: corehq/pillows/mappings/utils.py
Comment (sravfeyn) [reply corehq/pillows/mappings/utils.py:61]: What typo? I don't have a strong feeling either but I thought that that comments explains that multi_field is removed in ES7

File: corehq/pillows/mappings/utils.py
Comment (sravfeyn) [reply corehq/pillows/mappings/utils.py:47]: The idea is to mutate it recursively. Also It's a private function called from just `transform_for_es7`.

**Comment by sravfeyn** [comment]
@millerdev I merged the master. I tried rebasing as you suggested, but I lost some important changes, so I rolled back the rebase and made this PR as it was. I haven't been able to indentify why its' losing changes.

**Comment by sravfeyn** [comment]
The QA is complete for this PR on ES2 host without any issues. @millerdev This is the [QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1644) (it links to the test plan executed). It covers most of areas that use ES. Are there any other additional areas you would like to get tested for this to be QA complete?

**Comment by sravfeyn** [comment]
Mapping of commit to PR of what was already merged
- https://github.com/dimagi/commcare-hq/pull/28269 - https://github.com/dimagi/commcare-hq/pull/28270/commits/fc11624411cfc3509d3a91aeb57f94736366d011
- https://github.com/dimagi/commcare-hq/pull/28268 - https://github.com/dimagi/commcare-hq/pull/28270/commits/7aeafb2b39d56a93b13f2781b15dcc296497ecce
- https://github.com/dimagi/commcare-hq/pull/28267 - https://github.com/dimagi/commcare-hq/pull/28270/commits/3cb515f0c87349f839f432e2e7a66cf115022c86
- https://github.com/dimagi/commcare-hq/pull/28266 - https://github.com/dimagi/commcare-hq/pull/28270/commits/95b83dd4e82e35d0009ad08d7808a5237085fc79
- https://github.com/dimagi/commcare-hq/pull/28265 - https://github.com/dimagi/commcare-hq/pull/28270/commits/4bb34d527d9d000c1277ba4ce96ace2a90fff91e
- https://github.com/dimagi/commcare-hq/pull/28264 - https://github.com/dimagi/commcare-hq/pull/28270/commits/c6e7032e47c41af6b43279bb864f6a97c57eaa0a
- https://github.com/dimagi/commcare-hq/pull/28263 - https://github.com/dimagi/commcare-hq/pull/28270/commits/e7b7bf7ce1e3ef6f4f3a145d8effb47436c5d5df

File: corehq/pillows/mappings/utils.py
Comment (millerdev) [reply corehq/pillows/mappings/utils.py:61]: The phrase "just fields" confused me, and I thought that was a typo. I thought it should read

`# multi_field is replaced by text`

But maybe that was not the intent either?

File: corehq/pillows/mappings/utils.py
Comment (millerdev) [reply corehq/pillows/mappings/utils.py:47]: I see, and `transform_for_es7` makes a copy before sending it in here.

Previously [I advised you to remove a similar _copy_ call](https://github.com/dimagi/commcare-hq/pull/28267#discussion_r463839625), but I am now regretting that suggestion since it did not adhere to the best practice. The best practice is: _functions that return a value should not mutate any of their arguments_.

**Comment by millerdev** [comment]
@sravfeyn I made a new branch (rebased on master) with all of the already-merged changes excluded. I also used `fixup!` to eliminate lint other small cleanup commits. Can you tell me if it's equivalent to this one?

https://github.com/dimagi/commcare-hq/compare/dm/es7

**Comment by sravfeyn** [comment]
Thanks Daniel! Just these two lines missing in there, [this](https://github.com/dimagi/commcare-hq/pull/28270/files#diff-db6839e817d49823f5812ab21cf896faR15) and [this](https://github.com/dimagi/commcare-hq/pull/28270/files#diff-15a0a8683d5ad6a66fd387fb7a9667b7R30). Just imports. 

I am curious how you rebased without losing some of the commits.

**Comment by millerdev** [comment]
@sravfeyn thanks for looking it over. I have added those missing imports back into my branch. I am also working on some simplifications for commits like https://github.com/dimagi/commcare-hq/pull/28270/commits/c9c9c5d77d4719cdf61cb4830b65c7c5f9ad3d02 which hopefully will them easier to review. I will have you review the simplifications before I _fixup_ them into the commits they are intended to simplify.

> I am curious how you rebased without losing some of the commits.

The first step is to use `rebase --onto` to get a sequence of commits that does not include "merge master" commits like https://github.com/dimagi/commcare-hq/pull/28270/commits/123cde2d798eef4f457e23fc68e95dabdf590a00. This is done with multiple rebase commands of the form
```sh
# git rebase -i --onto <new_parent> <most_recent_merge_commit> <branch>
git rebase -i --onto c9b727acf 123cde2d7 dm/es7
```
The `<new_parent>` commit is the parent of `<most_recent_merge_commit>` that you you want to rebase onto. This will eliminate `<most_recent_merge_commit>` from the branch. Repeat this process for each "merge master" commit until there are no more merge commits in the branch.

Then do a `git fetch origin master`.

Next rebase the branch on `origin/master`. In this step many of the commits that were split out into other PRs are automatically "removed" (by git) since they have already been merged into master. There may be merge conflicts that need to be resolved. Resolving conflicts is the most tricky part. That is probably where I accidentally omitted the missing imports.

Finally, (this is not strictly part of the answer to your question) use rebase to squash (I mostly use `fixup` and `edit`) [lint](https://github.com/dimagi/commcare-hq/pull/28270/commits/c9b727acfaab1747615d01eacbc244ad03e2737a) and [small fixes](https://github.com/dimagi/commcare-hq/pull/28270/commits/d57f517c27fce6c0ce8421bbfaf5f4f8825174ed) into the larger logical commits on the branch.

**Comment by sravfeyn** [comment]
@millerdev Thanks! I have updated the commits of this branch, so now we have 9 commits that exclude the PRs that were merged.

File: corehq/apps/case_search/tests/test_filter_dsl.py
Comment (millerdev) [inline corehq/apps/case_search/tests/test_filter_dsl.py:329]: @sravfeyn the ES7 and ES2 versions of this query, while logically equivalent, are quite different. This is my reading of what they mean:

ES7: NOT (NOT property is set OR property  == "")
ES2: property is set AND NOT property == ""

I do see the ES7 version is produced by ``

https://github.com/dimagi/commcare-hq/blob/c5d65a8ff06e0b84819ea8378efa1cc5fb330500/corehq/apps/case_search/filter_dsl.py#L179-L185

and `case_property_missing`

https://github.com/dimagi/commcare-hq/blob/c5d65a8ff06e0b84819ea8378efa1cc5fb330500/corehq/apps/es/case_search.py#L245-L256

but I am not seeing how or where the ES2 version of this query is produced. Do you know where that's coming from or is there a bug in this test?

Comment (esoergel) [reply corehq/apps/case_search/tests/test_filter_dsl.py:329]: oooh, I think I know what's happening here.  This is testing `NOT(case_property_missing(...))`, and `case_property_missing` is itself sort of a negative check, so we end up with a double negative.  If we cared to, we could probably instead `case_property_exists` and define `case_property_missing` as the reverse.  That'd likely give smaller queries, if that even matters.

However, this reminded me of this code that came up in the migration to ES2, where we automatically simplify logic because ES appeared incapable of handling certain constructs:
https://github.com/dimagi/commcare-hq/blob/104765a7e1be3be5e45f7f157f197e115eb57a66/corehq/apps/es/filters.py#L49-L65
It looks like we're switching from `or` to `should`, which doesn't trigger that simplification.

Comment (millerdev) [reply corehq/apps/case_search/tests/test_filter_dsl.py:329]: Apparently ES7 does not have the same shortcoming as ES2 since it works even though that "simplification" is not applied to the ES7 query. I would vote to eliminate at least the first two branches (and possibly the third) of the `NOT` "simplification" in ES7. I think less automatic logic transformations will make it easier to troubleshoot queries.

However, this does make it a bit trickier write a function that automatically converts an ES7 query to an ES2 query (something I'm currently working on).

File: corehq/apps/api/tests/core_api.py
Comment (millerdev) [inline corehq/apps/api/tests/core_api.py:550]: The `query` submitted here uses ES2 syntax. I'm assuming the API logic does something to transform it to ES7 syntax. Does this mean our API will support ES2 query syntax indefinitely?

**Comment by millerdev** [comment]
I created a [new branch](https://github.com/dimagi/commcare-hq/compare/dm/es7-simplify) that attempts to simplify test changes by using a function to automatically transform from ES7 to ES2 query format (please DO NOT merge into this PR yet). This greatly simplifies the diffs (very simple example: [before](https://github.com/dimagi/commcare-hq/pull/28270/files#diff-a6cb819ac95f989426737a0329d27660) vs [after](https://github.com/dimagi/commcare-hq/compare/dm/es7-simplify#diff-a6cb819ac95f989426737a0329d27660), other more complex queries were even harder to compare). If we like this change I will squash those commits into the the original commits where the tests were modified.

There is a script in that branch [`check_es_queries.py`](https://github.com/dimagi/commcare-hq/compare/dm/es7-simplify#diff-9aabda9ec8a51425071b992312b85346), which can be run to verify that the transformations always produce the correct result. To run it, checkout the branch and run `python -m corehq.apps.es.tests.check_es_queries`. This script will be dropped if I merge the changes back into the original commits since it has served its purpose and has little/no enduring value.

**Comment by sravfeyn** [comment]
> I created a new branch that attempts to simplify test changes by using a function to automatically transform from ES7 to ES2 query format (please DO NOT merge into this PR yet). This greatly simplifies the diffs (very simple example: before vs after, other more complex queries were even harder to compare). If we like this change I will squash those commits into the the original commits where the tests were modified.

Thanks Daniel, I thought about this a bit and didn't take route. I took this approach of explicitly listing down the query for few reasons: 1. Explicit means we will know exactly what the query looks like 2. At a latter point we will have to remove ES2 queries and just keep ES7 queries, and 3.  Conversion script takes an extra effort that you are putting in. On the other hand the advantages of the approach you are doing is that it's less lines of code.

File: corehq/apps/api/tests/core_api.py
Comment (sravfeyn) [reply corehq/apps/api/tests/core_api.py:550]: No we don't plan to support the query at all. This sort of calls are coming from Data Export Tool, and one custom use-case, only these two we are supporting. Earlier, I had worked on [this PR](https://github.com/dimagi/commcare-hq/pull/28032) that discontinues support for random JSON ES queries from ES.  From that PR (which has been merged a while ago)
> Remove the ability to pass query part of elasticsearch query in public APIs. I made sure that this is not being used on prod by doing greps for these (grep 'api/v' | grep 'q=' and then grep 'api/v' | grep 'query=' over all proxy logs)
> Only accept two type of ES queries from _search GET param from known use cases. These I found based on grepping prod for _search param on api urls.

File: corehq/pillows/mappings/utils.py
Comment (sravfeyn) [reply corehq/pillows/mappings/utils.py:47]: Haha, exactly, generally I use copy whenever it's mutable arg.

File: corehq/pillows/mappings/utils.py
Comment (sravfeyn) [reply corehq/pillows/mappings/utils.py:61]: I will update.

File: corehq/apps/case_search/tests/test_filter_dsl.py
Comment (sravfeyn) [reply corehq/apps/case_search/tests/test_filter_dsl.py:329]: Yes, this is happening due what Ethan mentioned above. There is also a [test-case](https://github.com/dimagi/commcare-hq/pull/28270/files/c5d65a8ff06e0b84819ea8378efa1cc5fb330500#diff-2cdb698195be0262401dd2863151ca53R513) that I added where the expression is tested on the data that makes the expected behaviour explicit..

**Comment by millerdev** [comment]
> On the other hand the advantages of the approach you are doing is that it's less lines of code.

Additionally, and responding to your earlier points

1. We can easily see what the old and new queries look like by looking at the diff. This is easier than comparing large blocks of JSON that often do not fit on one screen.
2. When we remove ES2 queries the conversion is a tiny thing to be removed (two lines per query) and the ES7 query does not change indentation level.
3. The extra effort seemed worthwhile since it makes review easier and removal less disruptive.

**Comment by sravfeyn** [comment]
> Additionally, and responding to your earlier points

Alright Daniel. Anything to make the review easier I guess then. Let me know if you are planning to do that or if I should. (there might be few edge cases that make the conversions more tricky)

File: corehq/apps/es/queries.py
Comment (esoergel) [inline corehq/apps/es/queries.py:141]: this one is also ES2 compatible

File: corehq/apps/es/users.py
Comment (esoergel) [inline corehq/apps/es/users.py:35]: :+1:

**Review by esoergel — COMMENTED** [review]
I left some comments in-line, but the big piece of feedback is that I think all of the query syntax changes can be applied immediately, without needing to use different versions for ES2 and ES7.

It looks like the OR, AND, and NOT queries were actually [deprecated in ES2.0](https://www.elastic.co/guide/en/elasticsearch/reference/2.2/query-dsl-or-query.html), and so can be converted to `bool` queries immediately.  Same goes for a few others, see my in-line comments for those.

These changes would significantly simplify the rollout, and remove a notable variable to worry about when doing the actual switch to ES7.  We also wouldn't need to support both versions of the query builder while waiting on envs to upgrade off ES2.

**Comment by millerdev** [comment]
I squashed the new commits and rebased [dm/es7-simplify](https://github.com/dimagi/commcare-hq/compare/dm/es7-simplify) on latest master. I did not include https://github.com/dimagi/commcare-hq/pull/28270/commits/f32b71af1d665225b831051ea5dba8b87e3c2651 since I believe you had said it broke the build. The changes from [dm/es7-simplify](https://github.com/dimagi/commcare-hq/compare/dm/es7-simplify) should replace commits that were recently brought into this PR from [dm/es7](https://github.com/dimagi/commcare-hq/compare/dm/es7).

**Comment by sravfeyn** [comment]
@esoergel Thanks for the great insight that these queries are also compatible in ES2! I made a [separate PR](https://github.com/dimagi/commcare-hq/pull/28372/) that just updates these queries and can be merged without requiring ES7 upgrade.

**Review by millerdev — COMMENTED** [review]
I have updated [dm/es7](https://github.com/dimagi/commcare-hq/compare/dm/es7) with the suggestions and lint fixes from this PR, so this PR could be reset to that branch if you want.

File: corehq/apps/api/tests/user_resources.py
Comment (sravfeyn) [reply corehq/apps/api/tests/user_resources.py:11]: Yes, why not? This allows testing in ES7 and ES2 matrices.

File: corehq/apps/es/es_query.py
Comment (sravfeyn) [reply corehq/apps/es/es_query.py:125]: 👍

File: docker/hq-compose.yml
Comment (sravfeyn) [reply docker/hq-compose.yml:80]: I remember ES7 image having a different approach on volumes than earlier one, I don't recall exactly, but without this, the container wasn't running.

File: corehq/apps/es/es_query.py
Comment (sravfeyn) [reply corehq/apps/es/es_query.py]: Removed this `self.size(0).run().total` already does the same.

File: corehq/apps/api/tests/user_resources.py
Comment (millerdev) [reply corehq/apps/api/tests/user_resources.py:11]: Sorry, my question was not clear. I was asking if commit https://github.com/dimagi/commcare-hq/pull/28270/commits/c5d65a8ff06e0b84819ea8378efa1cc5fb330500 is still needed, or if it could be reverted? That is, is there still a merge conflict with staging?

File: corehq/apps/reports/analytics/esaccessors.py
Comment (esoergel) [inline corehq/apps/reports/analytics/esaccessors.py:366]: ooh, nice catch!

File: corehq/elastic.py
Comment (esoergel) [inline corehq/elastic.py:295]: What are the actual differences between these two versions?  Is the ES7 version for sure not available on ES2 as well?

It looks like `scan` is only called in one place anyways (`scroll_query`, just above this), and that is only called in one place (`ESQuery.scroll`), so I think there's actually a lot less variation than this stuff suggests.  In light of that, I'm betting this can be simplified significantly.  I see the docstring says

> This is a copy of elasticsearch.helpers.scan, except this function returns a ScanResult (which includes the total number of documents), and removes some options from scan that we aren't using.

I bet we can actually drop both of these functions and use that [`scan` library](https://elasticsearch-py.readthedocs.io/en/master/helpers.html?#scan) directly.  I think the only reason we reach in to the internals here is to get a count of documents without doing a second query, but looking at the places `.scroll()` is called, I don't actually see anywhere that uses that functionality - they all just treat it as an iterable, so that point is moot.  [These](https://github.com/dimagi/commcare-hq/blob/13228320848a504aa5e1b3d0488998e22e3e6c9d/corehq/apps/dump_reload/management/commands/dump_domain_data_raw.py#L89-L94) [two](https://github.com/dimagi/commcare-hq/blob/13228320848a504aa5e1b3d0488998e22e3e6c9d/corehq/apps/reports/standard/deployments.py#L434-L436) instances _could_ actually use that functionality, but I think I actually prefer a separate `count` query.

I think the possible approaches here are:

* Keep mostly as-is - I'd then simplify the code a bit so it's easier to follow, and try to unify as much as possible between the two versions if they really can't be unified
* Drop this entirely and make `ESQuery.scroll` call `elasticsearch.helpers.scan` and return a simple generator rather than a `ScanResult`

File: corehq/util/es/elasticsearch.py
Comment (esoergel) [inline corehq/util/es/elasticsearch.py:61]: shouldn't this be "1, 2, or 7"?  (Also, do we really still support ES1?)

File: corehq/pillows/mappings/utils.py
Comment (esoergel) [inline corehq/pillows/mappings/utils.py:17]: Is the aim here to replace the mapping file in a later PR after the migration?  It'd be a weird state to be in where we keep the mapping file in an out-of-date syntax and dynamically translate it to the "real" mapping.

File: corehq/util/es/elasticsearch.py
Comment (sravfeyn) [reply corehq/util/es/elasticsearch.py:61]: Right.

File: corehq/pillows/mappings/utils.py
Comment (sravfeyn) [reply corehq/pillows/mappings/utils.py:17]: Yes, once all envs are migrated, we just update the mapping files to the latest syntax and remove this tranformation

**Comment by millerdev** [comment]
@sravfeyn was it intentional that you added new commits to this branch? (https://github.com/dimagi/commcare-hq/pull/28270/commits/7da0873b4284c6ee4ae98f7c981b3d0cc48fa26e and following) It seems better to me to get this merged and then submit further work in subsequent PRs.

File: corehq/apps/es/es_query.py
Comment (esoergel) [inline corehq/apps/es/es_query.py]: oh nice, if this is only relevant to the scroll API, I like having it there more

File: corehq/elastic.py
Comment (esoergel) [inline corehq/elastic.py:240]: good catch - I totally missed this one

File: corehq/apps/es/es_query.py
Comment (esoergel) [inline corehq/apps/es/es_query.py:250]: thanks, this is great

**Review by esoergel — COMMENTED** [review]
I think the only open question I have left is the `_legacy_fields` business
https://github.com/dimagi/commcare-hq/pull/28270/files#r468726262

**Comment by sravfeyn** [comment]
@millerdev I have removed the reindex-tools from this PR now, so it's as it was earlier.

> I think the only open question I have left is the _legacy_fields business
https://github.com/dimagi/commcare-hq/pull/28270/files#r468726262

@esoergel I have removed that change now, I remember hitting some edge case, but if the build is passing now, we should be good without that

**Comment by millerdev** [comment]
Two more unused imports could be removed

- https://github.com/dimagi/commcare-hq/pull/28270/files/e052d340a5a2deded01eef9535c2e1f8f7017832#diff-2a8e5e1cac549af317ea4a90551e25eb
- https://github.com/dimagi/commcare-hq/pull/28270/files/e052d340a5a2deded01eef9535c2e1f8f7017832#diff-db6839e817d49823f5812ab21cf896fa

File: corehq/ex-submodules/pillowtop/es_utils.py
Comment (sravfeyn) [inline corehq/ex-submodules/pillowtop/es_utils.py:55]: @esoergel @millerdev FYI. On ES2 the default number of shards used to be 5 (so, all of our indexes have 5 shards unless customized via settings.ES_SETTINGS). ES7 default shards is just 1, so we will have to set this explicitly to 5 to have a similar scale up as existing clusters on all envs..

**Review by esoergel — APPROVED** [review]

**Review by millerdev — APPROVED** [review]

**Comment by sravfeyn** [comment]
The build of the reverse matrix is also pass from PR https://github.com/dimagi/commcare-hq/pull/28438 (all tests on ES7 pass). So merging this now.

**Comment by millerdev** [comment]
> The build of the reverse matrix is also pass from PR #28438 (all tests on ES7 pass). So merging this now.

@sravfeyn the last commit in this pr (https://github.com/dimagi/commcare-hq/pull/28270/commits/db1e1424924248c6979760be1068dc3db365ebbd) is the same as the last commit in #28438 (https://github.com/dimagi/commcare-hq/pull/28438/commits/db1e1424924248c6979760be1068dc3db365ebbd) which means that both PRs have the same set of commits. I think that explains why #28438 appears to be merged (it was automatically merged when #28270 was merged).

However, your comment implied that #28438 was setup to run tests against a "reverse matrix." I don't think this is true since both PRs contain identical commits. Am I missing something?

**Comment by sravfeyn** [comment]
Ah, thanks Daniel. I think this happened when I force-pushed on the other branch mistakenly, I will setup another build-only reverse matrix PR

**Comment by sravfeyn** [comment]
Hey 

@dimagi/team-commcare-hq  HQ now supports Elasticsearch 7 along with current version (2). If you are using docker locally, you will notice ES7 running on port 9200 and ES2 running on port 6200. You can use any version of ES on your HQ by updating your locasettings.ELASTICSEARCH_MAJOR_VERSION and locasettings.ELASTICSEARCH_PORT accordingly. To get your data indexed into ES7 you can run the regular preindex command.

Additionally, we also have a new travis test matrix to run all tests that have the [decorator](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/es/tests/utils.py#L51) `es_test` on both versions of Elasticsearch. So, if you are modifying any Elasticsearch related features, please make sure you add a test and decorate it with `@es_test` decorator. We plan to support both of these versions until all envs are upgraded.

**Pull request merged**
